### PR TITLE
Use UTC time fields provided by navitia 

### PR DESCRIPTION
**Pull request by pbougue**
The goal is to have all times in UTC inside Kirin (timezone-aware if possible)

It should be easier to read by commit as there are minor (unrelated) changes along the way.

JIRA: https://jira.kisio.org/browse/NAVP-1107

### CODE REVIEW

File: kirin/core/model.py [hunk 1]
```diff
@@ -96,42 +97,51 @@ class VehicleJourney(db.Model):
...
+    start_timestamp = db.Column(db.DateTime, nullable=False)  # ! USE get_start_timestamp() !
     db.Index('start_timestamp_idx', start_timestamp)
 
     db.UniqueConstraint(navitia_trip_id, start_timestamp,
                         name='vehicle_journey_navitia_trip_id_start_timestamp_idx')
 
-    def __init__(self, navitia_vj, local_circulation_date):
-        from kirin.utils import get_timezone
-
+    def __init__(self, navitia_vj, aware_since_dt, aware_until_dt):
```
Comment (pbougue) [inline kirin/core/model.py]: :dart: Core change is here (see code comment)
All other important changes are derived from this

File: kirin/gtfs_rt/model_maker.py [hunk 2]
```diff
@@ -196,66 +193,37 @@ def __repr__(self):
...
+                                  s=aware_since_dt,
+                                  u=aware_until_dt,
                                   ids=vj_ids
                                   ))
             record_internal_failure('duplicate vjs', contributor=self.contributor)
             return []
 
         nav_vj = navitia_vjs[0]
 
-        # Now we compute the real circulate_date of VJ from since, until and vj's first stop_time
```
Comment (pbougue) [inline kirin/gtfs_rt/model_maker.py]: This part (find and check circulate_date) is transfered into db.VehicleJourney builder:
https://github.com/CanalTP/kirin/pull/215/commits/15d03b1a7a86ab8575b9f725e36832b840aacd65#diff-d582ffa4339c008d83001df44608e3b8R127

File: tests/mock_navitia/vj_96231.py [hunk 3]
```diff
@@ -123,12 +123,12 @@
             ],
             "stop_times": [
                 {
-                    "arrival_time": "172100",
```
Comment (pbougue) [inline tests/mock_navitia/vj_96231.py]: Removing local "arrival_time" fields on that one to ensure we don't read it anymore.
Only this one as this fixture is heavily used and it's easy to add them back if needed later: maybe we should remove all of them as reading localtime is a bad idea...

File: tests/mock_navitia/vj_john.py [hunk 4]
```diff
@@ -520,12 +532,14 @@
             ],
             "stop_times": [
                 {
-                    "arrival_time": "123700",
+                    "arrival_time": "153700",
```
Comment (pbougue) [inline tests/mock_navitia/vj_john.py]: Change times to have a consistent response, start of VJ being between since and until (otherwise kirin now rejects it)

File: kirin/ire/model_maker.py [hunk 5]
```diff
@@ -130,12 +131,13 @@ def build(self, rt_update):
     def _get_vjs(self, xml_train):
         train_numbers = get_value(xml_train, 'NumeroTrain')
 
-        # to get the date of the vj we use the start/end of the vj + some tolerance
-        # since the ire data and navitia data might not be synchronized
-        vj_start = as_date(get_value(xml_train, 'OrigineTheoriqueTrain/DateHeureDepart'))
-        vj_end = as_date(get_value(xml_train, 'TerminusTheoriqueTrain/DateHeureTerminus'))
+        # stop_times are considered in Paris timezone for IRE
+        paris_tz = timezone('Europe/Paris')
```
Comment (pbougue) [inline kirin/ire/model_maker.py]: Using this TZ without computing it should be safe:
* Dead-alive code
* In SNCF's instance even London St Pancras is in Europe/Paris

File: tests/mock_navitia/vj_pass_midnight.py [hunk 6]
```diff
@@ -86,6 +86,8 @@
                 {
                     "arrival_time": "233000",
                     "departure_time": "233000",
+                    "utc_arrival_time": "033000",
```
Comment (pbougue) [inline tests/mock_navitia/vj_pass_midnight.py]: Not really pass midnight anymore (only in local TZ).
Maybe adding a test on UTC-pass-midnight would be nice.

File: tests/mock_navitia/vj_start_midnight.py [hunk 7]
```diff
@@ -86,6 +86,8 @@
                 {
                     "arrival_time": "000000",
                     "departure_time": "000000",
+                    "utc_arrival_time": "050000",
```
Comment (pbougue) [inline tests/mock_navitia/vj_start_midnight.py]: Not really start midnight anymore (only in local TZ).
Maybe adding a test on UTC-start-midnight would be nice.

File: kirin/core/handler.py [hunk 8]
```diff
@@ -139,30 +138,24 @@ def find_st_in_vj(st_id, vj_sts):
     return next((vj_st for vj_st in vj_sts if vj_st.get('stop_point', {}).get('id') == st_id), None)
 
 
-def convert_to_local_time(timezone, utc_time):
+def extract_str_utc_time(str_time):
     """
     Return local time according to UTC time and timezone
-    :param timezone: timezone info (type: datetime.tzinfo)
-    :param utc_time: UTC time (type: datetime.datetime)
+    :param str_time: UTC time (type: datetime.datetime)
```
Comment (kadhikari) [inline kirin/core/handler.py]: The function 'extract_str_utc_time' returns utc time or local time as indicated by comments ?

Comment (kadhikari) [inline kirin/core/handler.py]: The function 'extract_str_utc_time' returns utc time or local time as indicated by comments ?

File: kirin/core/model.py [hunk 9]
```diff
@@ -96,42 +99,50 @@ class VehicleJourney(db.Model):
...
+        """
+        Build a circulation (VJ on a given day) from:
+            * the navitia VJ (circulation times without a specific day)
+            * a datetime that's close but BEFORE the start of the circulation considered
+        This builds starting-datetime of the circulation to be the closest one after aware_since_dt.
```
Comment (mbounabi) [inline kirin/core/model.py]: Not understable. Also, the explanation about the extended datetime is already in the calling function (_get_navitia_vjs). Is it necessary to have it here again?

File: kirin/core/model.py [hunk 10]
```diff
@@ -96,42 +99,50 @@ class VehicleJourney(db.Model):
...
 
     db.UniqueConstraint(navitia_trip_id, start_timestamp,
                         name='vehicle_journey_navitia_trip_id_start_timestamp_idx')
 
-    def __init__(self, navitia_vj, local_circulation_date):
-        from kirin.utils import get_timezone
-
+    def __init__(self, navitia_vj, aware_since_dt, aware_until_dt):
+        """
+        Build a circulation (VJ on a given day) from:
```
Comment (mbounabi) [inline kirin/core/model.py]: Still not ok with the comments here:
- More 'Create' than 'Build'
- the navitia VJ retrieved has a specific day...

File: kirin/core/model.py [hunk 11]
```diff
@@ -96,42 +99,50 @@ class VehicleJourney(db.Model):
...
-        start_time = first_stop_time['arrival_time']
+        start_time = first_stop_time['utc_arrival_time']  # converted in datetime.time() in python wrapper
         if start_time is None:
-            start_time = first_stop_time['departure_time']
-        tzinfo = get_timezone(first_stop_time)
-        self.start_timestamp = tzinfo.localize(datetime.datetime.combine(local_circulation_date,
-                                                                         start_time)
-                                               ).astimezone(utc)
+            start_time = first_stop_time['utc_departure_time']  # converted in datetime.time() in python wrapper
+        utc_since = aware_since_dt.astimezone(utc)
```
Comment (mbounabi) [inline kirin/core/model.py]: At this point, 'aware_since_dt' should already have utc timezone info (and aware of it or it is wrongly named). 
Why is it necessary?

File: kirin/core/model.py [hunk 12]
```diff
@@ -96,42 +99,50 @@ class VehicleJourney(db.Model):
...
+        self.start_timestamp = utc.localize(datetime.datetime.combine(utc_since.date(), start_time))
+        if self.start_timestamp < utc_since:
+            self.start_timestamp += timedelta(days=1)
+        # simple consistency check (for now): the start timestamp must also be BEFORE aware_until_dt
+        utc_until = aware_until_dt.astimezone(utc)
+        if utc_until < self.start_timestamp:
+            msg = 'impossible to calculate the circulate date of vj: {} on period[{}, {}]'.format(
```
Comment (mbounabi) [inline kirin/core/model.py]: typo: circulation

**Comment by mbounabi** [comment]
I can't approve of this PR. It looks over-complicated for what it is supposed to do, hence the refinement that I suggested but never happened.

File: kirin/ire/model_maker.py [hunk 13]
```diff
@@ -130,12 +131,13 @@ def build(self, rt_update):
...
+        # to get the date of the vj we use the start/end of the vj
+        vj_start = paris_tz.localize(as_date(get_value(xml_train, 'OrigineTheoriqueTrain/DateHeureDepart')))
+        vj_end = paris_tz.localize(as_date(get_value(xml_train, 'TerminusTheoriqueTrain/DateHeureTerminus')))
 
-        return self._get_navitia_vjs(train_numbers, vj_start, vj_end)
+        return self._get_navitia_vjs(train_numbers, vj_start.astimezone(utc), vj_end.astimezone(utc))
```
Comment (TeXitoi) [inline kirin/ire/model_maker.py]: `astimezone(utc)` is useless, no?

File: kirin/core/model.py [hunk 14]
```diff
@@ -96,42 +99,50 @@ class VehicleJourney(db.Model):
...
-            start_time = first_stop_time['departure_time']
-        tzinfo = get_timezone(first_stop_time)
-        self.start_timestamp = tzinfo.localize(datetime.datetime.combine(local_circulation_date,
-                                                                         start_time)
-                                               ).astimezone(utc)
+            start_time = first_stop_time['utc_departure_time']  # converted in datetime.time() in python wrapper
+        utc_since = aware_since_dt.astimezone(utc)
+        self.start_timestamp = utc.localize(datetime.datetime.combine(utc_since.date(), start_time))
+        if self.start_timestamp < utc_since:
+            self.start_timestamp += timedelta(days=1)
```
Comment (TeXitoi) [inline kirin/core/model.py]: This need a comment.

File: kirin/core/model.py [hunk 15]
```diff
@@ -96,42 +99,50 @@ class VehicleJourney(db.Model):
...
+        # simple consistency check (for now): the start timestamp must also be BEFORE aware_until_dt
```
Comment (mbounabi) [inline kirin/core/model.py]: Is there a possibility for this to happen? 
These parameters are calculated from the pdp_list in the COTS feed. At this point, we should already have determined if the COTS feed is valid or not. To be verified and maybe done in another PR.
